internal/server/singleprocess/service_task.go

```diff
@@ -0,0 +1,123 @@
+package singleprocess
+
+import (
+	"context"
+
+	"github.com/golang/protobuf/ptypes/empty"
+
+	pb "github.com/hashicorp/waypoint/pkg/server/gen"
+	serverptypes "github.com/hashicorp/waypoint/pkg/server/ptypes"
+)
+
+func (s *service) UpsertTask(
+	ctx context.Context,
+	req *pb.UpsertTaskRequest,
+) (*pb.UpsertTaskResponse, error) {
+	if err := serverptypes.ValidateUpsertTaskRequest(req); err != nil {
+		return nil, err
+	}
+
+	result := req.Task
+	if err := s.state.TaskPut(result); err != nil {
+		return nil, err
+	}
+
+	return &pb.UpsertTaskResponse{Task: result}, nil
+}
+
+// GetTask returns a Task based on ID
+func (s *service) GetTask(
+	ctx context.Context,
+	req *pb.GetTaskRequest,
+) (*pb.GetTaskResponse, error) {
+	if err := serverptypes.ValidateGetTaskRequest(req); err != nil {
+		return nil, err
+	}
+
+	t, err := s.state.TaskGet(req.Ref)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the Start, Run, and Stop jobs
+	resp, err := s.getJobsByTaskRef(ctx, t)
```

Comment:
I think we should hoist this logic into the state package and introduce an API like `TaskGetWithJobs`. That way, we can get all the jobs under the same memdb read-lock for consistency.